### ISSUE TITLE
CSS shows up twice when Critical CSS Enabled

### ISSUE THREAD

**Issue by brightonmike** [magento/magento2#26498]
### Preconditions (*)

1. 2.3.4
2.

### Steps to reproduce (*)

1. Enable `use_css_critical_path`

### Expected result (*)

1. The stylesheets are only loaded once

### Actual result (*)

1. `styles-m.css`, `styles-l.css` and `print.css` styles show up multiple times in developer tools.

I have traced the problem to the following method:
`beforeSendResponse` in `module-theme/Controller/Result/AsyncCssPlugin.php`

This line appears to be re-inserting the original style tags:

```
$content = str_replace('</body', implode("\n", $cssMatches) . "\n</body", $content);
```

If you comment out this line, the HTML is as expected, with each style element existing only once and preventing double download of stylesheets.

I think there's some confusion over the implementation of the following technique:

```
<link rel="preload" as="style" media="all" onload="this.onload=null;this.rel='stylesheet'"href="...en_US/css/styles-m.css">
```

I would expect that the purpose of the onload method is that rather than needing one `<link>` to preload and a separate `<link>` at the bottom of the HTML to then achieve the affect of asynchronously loading the CSS, this behaviour can be achieved with just one `<link>` element. After `onload` event has fired, the `rel` attribute swaps and applies the CSS.

**Comment by brightonmike** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @brightonmike. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @brightonmike, here is your [Magento instance](https://i-26498-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-26498-2-4-develop.instances.magento-community.engineering/admin_95bd](https://i-26498-2-4-develop.instances.magento-community.engineering/admin_95bd)
Login: `b4063ce2` Password: `32017648ea26`
Instance will be terminated in up to 3 hours.

**Comment by brightonmike** [comment]
![Screen Shot 2020-01-22 at 15 19 11](https://user-images.githubusercontent.com/2071150/72906740-9242e600-3d2a-11ea-8563-15ec25bc1b67.png)


Please see screenshot. As you will see the styles are showing up twice per style sheet.

For myself personally, this is causing problems debugging FE code. It looks like this issue is to force the specificity of the stylesheets over the critical inlined CSS.

Couple of possible work arounds:

- Remove the `onload` event. The first style tags are simply for preload.
- Or, remove the second style tags and move the first style tags below the inlined CSS.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-30771` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by donatopirolo** [comment]
I have encountered the same problem with magento 2.3.5. Has anyone managed to solve it?

**Comment by davidwindell** [comment]
@donatopirolo  there's already a working merge request linked to this issue

**Comment by magento-engcom-team** [comment]
Hi @brightonmike. Thank you for your report. 
The issue has been fixed in magento/magento2#28480 by @mrtuvn in 2.4-develop branch
Related commit(s):
 - [bacb8172fb9d2df6b93adacd5cf115b725b45da9](https://github.com/magento/magento2/commit/bacb8172fb9d2df6b93adacd5cf115b725b45da9)

The fix will be available with the upcoming 2.4.1 release.

**Comment by HenKun** [comment]
1. Just for reference: The fix is not included in 2.4.1
2. Similar to `JsFooterPlugin` the sort order of `AsyncCssPlugin` need to be negative. Else, the changes made by the plugin are not included in the cached result, when using FPC. So when the cache is not hit, CSS is processed correctly, however, if the cache is hit, the plugin is neither called nor are the changes already included in the cached response. The FPC uses a plugin on the same type (e.g. `Magento\PageCache\Model\Controller\Result\BuiltinPlugin` on `Magento\Framework\Controller\ResultInterface` which is a base of `Magento\Framework\View\Result\Layout`). So the `AsyncCssPlugin` needs to be called before the caching plugin. @mrtuvn Should I create a separate ticket for this?

**Comment by mrtuvn** [comment]
Yes this update only available in 2.4-develop and maybe available in next version

**Comment by rafczow** [comment]
Thanks @HenKun! After change of sort order to negative on AsyncCssPlugin it finally works as intended after cache hit.
This should be included with merged fix as without it changes are worthless anyway for any store that uses caching which basically means every production store.

**Comment by hostep** [comment]
@rafczow, @HenKun, the change is coming in Magento 2.4.3, will get released on August 10, fixed in https://github.com/magento/magento2/pull/30871

**Comment by mrtuvn** [comment]
Hi i just curiously anything else broken function in pre-preview 2.4.3 ? Not have a change to test that
